lib/grizzly/zwave_firmware.ex

```diff
@@ -0,0 +1,155 @@
+
+    case applicable_firmware_info(firmware_info_list, current_firmware) do
+      %{path: path} ->
+        dir = Application.app_dir(options.zwave_support, options.zwave_firmware_dir)
+        full_path = Path.join(dir, path)
+
+        if File.exists?(full_path) do
+          {:ok, full_path}
+        else
+          failure(options, "Firmware image file #{full_path} does not exist")
+          {:error, :fw_image_file_not_found}
+        end
+
+      nil ->
+        Logger.info(
+          "[Grizzly] No applicable firmware upgrade found in #{inspect(firmware_info_list)} for current firmware #{inspect(current_firmware)}"
+        )
+
+        {:error, :fw_image_file_not_found}
+    end
+  end
+
+  defp applicable_firmware_info(
+         firmware_info_list,
+         %{chip_type: chip_type, version: current_version} = _current_firmware
+       ) do
+    Enum.find(firmware_info_list, fn %{
+                                       chip_type: target_chip_type,
+                                       version: target_version
+                                     } ->
+      target_chip_type == chip_type and
+        upgrade_valid?(current_version, target_version)
+    end)
+  end
+
+  # {:ok, %{chip_type: 7, version: "7.15.2"}}
+  defp current_firmware(options) do
+    with {answer, 0} <- System.cmd(options.zw_programmer_path, ["-s", options.serial_port, "-t"]),
+         :ok <- check_healthy(answer),
+         {:ok, current_version} <- extract_current_version(answer),
+         {:ok, chip_type} <- extract_chip_type(answer, options) do
+      {:ok, %{chip_type: chip_type, version: current_version}}
+    else
+      error ->
+        error
+    end
+  end
+
+  defp check_healthy(answer),
+    do: if(String.contains?(answer, "Serial Init failed"), do: {:error, :not_healthy}, else: :ok)
+
+  # "Using serial device /dev/ttyS4\nConnected to Serial device: OK\nSerial version: 9, Chip type: 7, Chip version: 0, SDK: 7.15.02, ..."
+  defp extract_current_version(answer) do
+    Logger.info("[Grizzly] Extracting current version from #{inspect(answer)}")
+    regex = ~r/.*SDK:\s*(?<major>\d+)\.(?<minor>\d+)\.(?<patch>\d+)/m
+
+    Regex.named_captures(regex, answer)
+    |> parse_version_capture()
+  end
+
+  defp extract_chip_type(answer, options) do
+    regex = ~r/Chip type: (?<chip_type>\d+)/m
+
+    case Regex.named_captures(regex, answer) do
+      %{"chip_type" => chip_type_s} ->
+        {chip_type, ""} = Integer.parse(chip_type_s)
+        {:ok, chip_type}
+
+      nil ->
+        :ok = failure(options, "Unknown chip type")
+        {:error, :unknown_chip_type}
+    end
+  end
+
+  # Can not parse using Version.parse/1 because it fails on "7.15.02" etc.
+  defp parse_version_capture(%{"major" => major_s, "minor" => minor_s, "patch" => patch_s}) do
+    {major, ""} = Integer.parse(major_s)
+    {minor, ""} = Integer.parse(minor_s)
+    {patch, ""} = Integer.parse(patch_s)
+    {:ok, %Version{major: major, minor: minor, patch: patch}}
```

Comment:
See https://hexdocs.pm/elixir/Version.html#module-struct. We'll have to do something different since creating a `Version` struct manually isn't supported. I'm vaguely aware that there was a problem caused by people doing this in the past. I'm less sure what the problem is now, but the docs are pretty specific about this.